### PR TITLE
Extract depth algorithm and improve performance.

### PR DESCRIPTION
**Pull request by alumi**
#### Changes

- Extracted and renamed `cljam.algo.pileup.pileup/pileup` function to `cljam.algo.depth/lazy-depth`.
- Added eager-version `cljam.algo.depth/depth` optimized for performance.
- Added some utility functions to manipulate region.

#### Benchmark

Computing depth for `chr1:1-248956422`  BAM file containing about 3500000 alignments in the region.

##### Environment
- CPU
  - Corei7-6700K
  - 4.00GHz
  - 4 cores
  - 8 threads
- JVM
  - HotSpot 1.8.0_131
  - MaxHeapSize 16861102080
- criterium 0.4.4

```clojure
(with-open [r (sam/reader "BAM_FILE")]
  (count (depth/FN r {:chr "chr1", :start 1, :end 248956422} {:n-threads 8, OPTS})))
```

##### Results

|fn|time|
|:-|-:|
|lazy-depth|83934ms|
|depth|7848ms|
|depth unchecked|6500ms|

The eager version is >10x faster.

### CODE REVIEW

File: src/cljam/algo/depth.clj [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+(defn depth
+  "Calculate depth of coverage eagerly. Returns a seq of depth for range [start, end].
+  Requires a `cljam.io.bam.reader.BAMReader` instance and region.
+  If start and end are not supplied, piles whole range up.
+  Note that CIGAR code in alignemnts are ignored and only start/end positions are used."
+  [bam-reader {:keys [chr start end] :or {start 1 end Long/MAX_VALUE}}
+   & [{:keys [step unchecked? n-threads] :or {step default-step unchecked? false n-threads 1}}]]
+  {:pre [chr start end (pos? start) (pos? end) (<= start end)]}
+  (when-let [{:keys [len] :as r} (sam-util/ref-by-name (sam/read-refs bam-reader) chr)]
```
Comment (totakke) [inline src/cljam/algo/depth.clj]: `r` is never used.

File: src/cljam/algo/pileup.clj [hunk 2]
```diff
@@ -1,29 +1,11 @@
...
-            [cljam.algo.pileup.pileup :as plp]
+  (:require [cljam.io.sam :as sam]
             [cljam.algo.pileup.mpileup :as mplp]))
 
-(defn first-pos
-  "Returns a position of first alignment in left-right, or nil."
-  [bam-reader region]
-  (plp/first-pos bam-reader region))
-
 (def ^:private default-pileup-option
```
Comment (totakke) [inline src/cljam/algo/pileup.clj]: `default-pileup-option` is no longer needed.

**Comment by alumi** [comment]
Thank you for reviewing! I've just removed the unused vars.

**Comment by totakke** [comment]
Thanks

**Pull request merged**